swdb_2018_tools/GLM_helpers.py

```diff
@@ -0,0 +1,98 @@
+import statsmodels.api as sm
+import numpy as np
+import pandas as pd
+import keras
+from keras import backend as K
+from keras.models import Model
+from keras.layers import Input
+from keras import Sequential
+from keras.layers import Dense, Lambda
+
+# Inputs:
+# stimulus: video shown to the mouse in some form (M pixels x T time points)
+# spikes: spikes recorded from each neuron in some form (N neurons x T time points)
+# d_stim: length of stimulus filters in samples
+# d_spk: length of spike filters in samples
+# f: string for the firing rate function used. options are {'exp'}
+# priors: TODO
+# L1: regularization parameter for sparse synapses TODO
+
+# Returns:
+# GLM network model with parameters fit
+
+def GLM_network_fit(stimulus,spikes,d_stim, d_spk,bin_len,f='exp',priors=None,L1=None):
+    N = spikes.shape[0]
+    M = stimulus.shape[0]
+    F = np.empty((N,M,d_stim)) # stimulus filters
+    W = np.empty((N,N,d_spk))  # spike train filters
+    b = np.empty((N,)) # biases
+    fs = {'exp':K.exp}
+    Xdsn = construct_GLM_mat(stimulus,spikes, d_stim, d_spk)
+    for i in range(N):
+        y = spikes[i,max(d_stim,d_spk):]
+        # construct GLM model and return fit
+        model = Sequential()
+        model.add(Dense(1,input_dim = Xdsn.shape[1],use_bias=True))
+        model.add(Lambda(lambda x: fs[f](x)*bin_len))
+        model.compile(loss = 'poisson',optimizer = keras.optimizers.adam(lr=2e-1))
+        model.fit(x=Xdsn,y=y,epochs=150,batch_size = 1000000,verbose=0)
+        p = model.get_weights()[0]
+        F[i,:,:] = p[:M*d_stim].reshape((M,d_stim))
+        W[i,:,:] = p[M*d_stim:].reshape((N,d_spk))
+        b[i] = model.get_weights()[1]
+    return (F,W,b)
+
+
+# Inputs:
+# data_set: EphysObservatory data_set structure
+# bin_len: duration of a bin in seconds
+# t_start: time to start counting spikes
+# t_final: time to stop counting spikes
+# probes: list of strings of probe names to be used (default is all probes)
+# regions: list of strings of brain regions to be used (default is all regions)
+def bin_spikes(data_set,bin_len,t_start,t_final,probes=None,regions=None):
+    if probes is None:
+        probes = data_set.probe_list
+    if regions is None:
+        regions = data_set.unit_df.structure.unique()
+    
+    #gather cells from desired regions and probes into cell_table
+    use_cells = False
+    for probe in probes:
+        for region in regions:
+            use_cells |= (data_set.unit_df.probe==probe) & (data_set.unit_df.structure==region)
+    cell_table = data_set.unit_df[use_cells]
+    
+    N = len(cell_table)     #number of cells
+    T = int(np.floor((t_final-t_start)/bin_len)) #number of time bins
+    binned_spikes = np.zeros((N,T)) # binned_spikes[i,j] is the number of spikes from neuron i in time bin j
+
+    #for each cell in the table, add each spike to the appropriate bin
+    i = 0
+    for z,cell in cell_table.iterrows(): 
```

Comment:
you might be able to improve the performance of this section using numpy's searchsorted function.